MIFOSAC-552 Implement Settings Step for New Recurring Deposits Account Flow

TheKalpeshPawar opened the pull request:
[Jira-#MIFOSAC-552](https://mifosforge.jira.com/browse/MIFOSAC-552)

|||
|---|---|
|Updated Video|Later updated video for currency Symbol|
|<video src="https://github.com/user-attachments/assets/f9d4f621-93c9-45fb-b450-909592daf906" />|<video src="https://github.com/user-attachments/assets/3b19d3a0-b943-49d4-8ab9-ad365df7eb6c" /> |


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full recurring deposit flow: multi-step account creation with detailed settings, loading/error dialogue with retry, currency-aware amount formatting, and template loading by product.
  * Navigation now requires a client selection (clientId) when starting a recurring account.
  * Offline detection and retry support.
  * New localised strings for recurring deposit screens.

* **Bug Fixes**
  * Repository method names clarified for recurring-account template access.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TheKalpeshPawar committed:
feat: enhance recurring deposit account settings and UI components

TheKalpeshPawar committed:
feat: enhance recurring deposit account settings and UI components

TheKalpeshPawar committed:
feat: add keyboard options for number input in recurring deposit settings

TheKalpeshPawar committed:
feat: add keyboard options for number input in recurring deposit settings

TheKalpeshPawar committed:
fix: add missing commas and improve formatting in recurring deposit account files

TheKalpeshPawar committed:
spotless fixes

TheKalpeshPawar committed:
feat: implement loading state indicator and refactor dialog state management in recurring account

TheKalpeshPawar committed:
feat: add RecurringDepositAccountDialogBox to RecurringAccountScreen

TheKalpeshPawar committed:
fix: remove unnecessary blank line in RecurringAccountViewModel

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountScreen.kt:
```diff
@@ -110,3 +116,24 @@ private fun RecurringAccountScaffold(
+                },
+            )
+        }
+        RecurringAccountState.DialogState.Loading -> {
```
The Loading State should be a `ScreenState` and not a `DialogState`.

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt:
```diff
@@ -9,21 +9,184 @@
+                // Must be removed from here after the implementation of detail screen is finished.
+                loadTemplateByProduct()
+            } else {
+                setErrorState("No internet connection")
```
Hardcoded string can be moved to `strings.xml` and use `getString()` here

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt:
```diff
@@ -9,21 +9,184 @@
+                // date in dd MM yyyy format.
+                submittedOnDate = s.recurringDepositAccountDetail.submittedOnDate,
+            )
+            recurringAccountRepository.createRecurringDepositAccount(payload).collect { dataState ->
```
Before making a network call, check if the user is online or not. So include the logic from observe network here too.

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt:
```diff
@@ -38,33 +201,296 @@ class RecurringAccountViewModel : BaseViewModel<
+                        sendEvent(RecurringAccountEvent.NavigateBack)
+                    }
+                    RecurringAccountAction.RecurringAccountSettingsAction.OnNextPress -> {
+                        if (state.currentStep < state.totalSteps) {
```
This part is same for the `RecurringAccountAction.NextStep` action as  well, and can be extracted into a function like so: 
```
private fun moveToNextStep() {
    val current = state.currentStep
    if (current < state.totalSteps) {
        mutableStateFlow.update {
            it.copy(
                currentStep = current + 1,
            )
        }
    } else {
          sendEvent(SavingsAccountEvent.Finish)
    }
}
```

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/pages/SettingsPage.kt:
```diff
@@ -9,23 +9,340 @@
+            onFirstBtnClick = { onAction(RecurringAccountAction.RecurringAccountSettingsAction.OnBackPress) },
+            onSecondBtnClick = { onAction(RecurringAccountAction.RecurringAccountSettingsAction.OnNextPress) },
+            isButtonIconVisible = true,
+            isSecondButtonEnabled = settingsState.preMatureClosure.penalInterest.isNotBlank() &&
```
Could extract this validation logic into the viewModel and use a state to represent this, something like: 

```
val isTermsNextEnabled = isDetailsNextEnabled &&
            currencyIndex != -1 &&
            interestCalcIndex != -1 &&
            interestPostingPeriodIndex != -1 &&
            interestCompPeriodIndex != -1
```

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt:
```diff
@@ -9,21 +9,184 @@
     >(RecurringAccountState()) {
 
+    init {
+        observeNetwork()
```
I believe `observerNetwork()` is not really required. Instead only check network status right before making an api call or network request. So for each request include this logic before making the network request like so: 

```
private fun loadClientTemplate() = viewModelScope.launch {
        val online = networkMonitor.isOnline.first()
        if (online) {
            getClientTemplateUseCase().collect { result ->
                sendAction(SavingsAccountAction.Internal.OnReceivingClientTemplate(result))
            }
        } else {
            mutableStateFlow.update {
                it.copy(
                    screenState = SavingsAccountState.ScreenState.NetworkError,
                )
            }
        }
    }

    private fun loadSavingsProductTemplate() = viewModelScope.launch {
        val online = networkMonitor.isOnline.first()
        if (online) {
            getSavingsProductTemplateUseCase().collect { result ->
                sendAction(SavingsAccountAction.Internal.OnReceivingSavingsProductTemplate(result))
            }
        } else {
            mutableStateFlow.update {
                it.copy(
                    screenState = SavingsAccountState.ScreenState.NetworkError,
                )
            }
        }
    }
```
For reference you can check the NewLoanAccount pr's and the NewSavingsAccount pr's that are already merged.

sam-arth07 reviewed (commented):
Suggestion: For the viewModel, you can take ref from how the vm is written in `loanAccount` or `savingsAccount`

Some additional changes: 
1. > Checkbox options dynamically influence field visibility (penal interest fields should only appear if checked).

Currently the field are visible even when Apply Penal Interest is unchecked.

2. The spacing looks uneven.
<img width="250" height="500" alt="Image" src="https://github.com/user-attachments/assets/2815ba8f-9570-4eca-8e8d-f8af06e1f87e" />

UI mockup: 

<img width="250" height="500" alt="Image" src="https://github.com/user-attachments/assets/037b637c-0de1-4095-90ac-4d41f4108f68" />

3. Same here: 
<img width="250" height="500" alt="Image" src="https://github.com/user-attachments/assets/6aeb82d9-e5b8-4c0f-a1b6-b98c6f1f9cd9" />

UI mockup: 

<img width="250" height="500" alt="Image" src="https://github.com/user-attachments/assets/f83db0dd-daae-4d92-b4a0-d140171d122d" />

biplab1 commented on core/data/src/commonMain/kotlin/com/mifos/core/data/repository/RecurringAccountRepository.kt:
```diff
@@ -17,9 +17,9 @@ import kotlinx.coroutines.flow.Flow
+    fun getRecuttingAccountTemplate(): Flow<DataState<RecurringDepositAccountTemplate>>
 
-    fun getRecuttingAccountRepositoryBtProduct(
+    fun getRecuttingAccountTemplateByProduct(
```
It should `Recurring` instead of `Recutting`

biplab1 commented on core/data/src/commonMain/kotlin/com/mifos/core/data/repositoryImp/RecurringAccountRepositoryImp.kt:
```diff
@@ -21,12 +21,12 @@ import kotlinx.coroutines.flow.Flow
     }
 
-    override fun getRecuttingAccountRepositoryBtProduct(
+    override fun getRecuttingAccountTemplateByProduct(
```
Same here, please review.

TheKalpeshPawar committed:
refactor: rename dialogState to screenState and update related logic in RecurringAccount components

TheKalpeshPawar committed:
refactor: rename dialogState to screenState and update related logic in RecurringAccount components

TheKalpeshPawar committed:
fix: clean up formatting and remove unnecessary blank lines in RecurringAccountViewModel and SettingsPage

TheKalpeshPawar committed:
feat: improve amount formatting logic and enhance UI layout in RecurringAccount settings

TheKalpeshPawar committed:
fix: remove duplicate import of RecurringDepositModule in KoinModules

TheKalpeshPawar committed:
feat: enhance amount formatting and validate network connectivity in RecurringAccount

TheKalpeshPawar committed:
feat: rename network observation method and enhance connectivity checks in RecurringAccountViewModel

TheKalpeshPawar committed:
refactor: clean up code formatting and improve readability in DataManagerModule and RecurringAccountViewModel

biplab1 commented on core/model/src/commonMain/kotlin/com/mifos/core/model/objects/payloads/RecurringDepositAccountPayload.kt:
```diff
@@ -13,9 +13,9 @@ import kotlinx.serialization.Serializable
 
 @Serializable
 data class RecurringDepositAccountPayload(
+    //    val charges: List<Any>? = null,
```
Can you tell me why this has been commented out?

TheKalpeshPawar replied:
I couldn't find a representation of that parameter in the api call.
It sends an empty array or object.

I also couldn't find its use.
So, in future, if someone does find its use or the proper representation of the data type, they can update the code.

biplab1 replied:
It is better to add a TODO comment for future reference.

TheKalpeshPawar replied:
Okay sir. I will keep this in mind.

biplab1 commented on feature/recurringDeposit/src/commonMain/composeResources/values/string.xml:
Please follow the naming convention for a string resource used in other feature modules. The name should be prefixed with the name of the feature. Check other feature modules where it has been implemented appropriately. This avoids wrong usage of strings elsewhere.

TheKalpeshPawar replied:
I fixed this issue now.
Actually, I had AI extract hardcoded strings for me from the files 😅 so, I did not notice the naming convention was not followed.

biplab1 replied:
You could tell the AI to follow the naming convention as in other modules. I am sure it will follow the instructions but one needs to steer its direction when it chooses a wrong direction.

TheKalpeshPawar replied:
Sir, I fixed that issue. I will be careful next time.

TheKalpeshPawar committed:
feat: rename string resources for recurring deposit feature

Arinyadav1 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountScreen.kt:
```diff
@@ -64,39 +68,40 @@ private fun RecurringAccountScaffold(
     ) { paddingValues ->
-        if (state.dialogState == null) {
+        if (state.screenState == null) {
             MifosStepper(
```
Why not use here breadcrumb, Please Add breadcrumb before calling Stepper

TheKalpeshPawar replied:
I did not implement it someone else did. I also did not notice though.

Arinyadav1 replied:
Okay so please fix it

TheKalpeshPawar replied:
I have added the MifosBreadcrumbBar.

Arinyadav1 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountScreen.kt:
```diff
@@ -33,13 +34,14 @@ import com.mifos.feature.recurringDeposit.newRecurringDepositAccount.pages.Inter
+import org.koin.compose.viewmodel.koinViewModel
 
 @Composable
 internal fun RecurringAccountScreen(
```
Take reference from `create new loan` and `create new saving` account for handle proper screen state where you also find how breadcrumb use.

TheKalpeshPawar committed:
Merge branch 'development' into settings-recurring-deposit-account

TheKalpeshPawar committed:
feat: update navigation for recurring account flow and improve layout in RecurringAccountScreen

biplab1 commented on version.txt:
This file shouldn't be part of this PR. Please review.

TheKalpeshPawar replied:
I don't know what it is for. It was already present.
Maybe it should be added to .gitignore.

biplab1 commented:
During typing the `Recurring Deposit Account`, the currency symbol (a dollar sign) seems to be appear by itself. Also, the cursor is is showing before the first last digit instead of showing after it. Please review. Screenshot attached for reference.

<img width="178" height="348" alt="image" src="https://github.com/user-attachments/assets/628d7afc-ba6d-44cd-9d74-ea03595a14a3" />

biplab1 commented on feature/recurringDeposit/src/commonMain/composeResources/values/feature_recurringDeposit_string.xml:
```diff
@@ -0,0 +1,39 @@
+    See https://github.com/openMF/android-client/blob/master/LICENSE.md
+-->
+<resources>
+    <string name="feature_recurringDeposit_step_details">Details</string>
```
This is the convention followed for modules names with two words. 

```suggestion
    <string name="feature_recurring_deposit_step_details">Details</string>
```

TheKalpeshPawar replied:
- DOne
-

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/pages/SettingsPage.kt:
```diff
@@ -9,23 +9,351 @@
+            }
+        }
+
+        val isNextButtonActive = settingsState.preMatureClosure.penalInterest.isNotBlank() &&
```
Instead of making all these checks in the UI, which makes it harder to read , move them into the view Model. They can be represented by a state and call that state here like: 
`state.isSettingsNextEnabled`

TheKalpeshPawar replied:
Done

sam-arth07 commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/pages/SettingsPage.kt:
```diff
@@ -9,23 +9,351 @@
+                )
+                MifosTextFieldDropdown(
+                    value = if (settingsState.preMatureClosure.interestPeriodIndex != -1) {
+                        state.recurringDepositAccountTemplate.preClosurePenalInterestOnTypeOptions?.get(settingsState.preMatureClosure.interestPeriodIndex)?.value ?: ""
```
```suggestion
                        state.recurringDepositAccountTemplate.preClosurePenalInterestOnTypeOptions?.getOrNull(settingsState.preMatureClosure.interestPeriodIndex)?.value.orEmpty()
```

TheKalpeshPawar replied:
Done

sam-arth07 reviewed (commented):
<img width="250" height="500" alt="image" src="https://github.com/user-attachments/assets/5a5ab6dd-b3db-4ae1-8b1a-e2afb1f71c58" />

There seems to be an extra spacer between `Period` and 'Min balance for interest calculation.'

Also remove the version file from this pr, as suggested by @biplab1 .

TheKalpeshPawar committed:
feat: rename string resources for recurring deposit feature and update UI components

TheKalpeshPawar committed:
Merge branch 'development' into settings-recurring-deposit-account-merge

# Conflicts:
#	cmp-navigation/src/commonMain/kotlin/cmp/navigation/di/KoinModules.kt
#	feature/client/src/commonMain/kotlin/com/mifos/feature/client/clientApplyNewApplications/ClientApplyNewApplicationsScreen.kt
#	feature/client/src/commonMain/kotlin/com/mifos/feature/client/navigation/ClientNavigation.kt
#	feature/recurringDeposit/src/commonMain/composeResources/values/string.xml
#	feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountRoute.kt
#	feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountScreen.kt
#	feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt

TheKalpeshPawar committed:
Refactor string resources for recurring deposit feature: move hardcoded strings to feature_recurring_deposit_string.xml and update references in ViewModel and page composables.

TheKalpeshPawar committed:
Refactor string resources and UI spacing for recurring deposit feature
Move hardcoded strings from string.xml to feature_recurring_deposit_string.xml, following the naming convention "feature_recurring_deposit_" followed by lowercase with underscores.
Update RecurringAccountViewModel.kt to replace hardcoded error messages ("Deposit amount is required", "Lock-in period frequency is required", "Recurring frequency is required") with references to the new string resources.
Update page composables (DetailsPage.kt, TermsPage.kt, InterestPage.kt, ChargesPage.kt) to use string resources instead of hardcoded strings for titles and buttons.
Adjust spacing in SettingsPage.kt by wrapping the leadingIcon Text in a Box with Modifier.padding(end = 4.dp) to decrease the gap between the currency symbol icon and the text input in both MifosOutlinedTextField instances (around lines 139 and 329).
Verify that all string resources used in the updated files are present in feature_recurring_deposit_string.xml.

revanthkumarJ commented on feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt:
```diff
@@ -254,17 +677,47 @@ data class RecurringAccountSettingsState(
+    object OnNextPress : RecurringAccountAction()
+    object Finish : RecurringAccountAction()
+    data object Retry : RecurringAccountAction()
+
```
can you merge all actions into 1 i think while merging they got mixed

TheKalpeshPawar replied:
Done

TheKalpeshPawar committed:
Reduce leading icon spacing in MifosOutlinedTextField and tidy recurring-deposit UI

Adjusted the leading icon spacing inside `core/designsystem/src/commonMain/kotlin/com/mifos/core/designsystem/component/MifosOutlinedTextField.kt` to bring the icon closer to the text input for a more compact, consistent form appearance. Updated recurring deposit pages to use the revised text field behavior and to align paddings and labels: `feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/pages/DetailsPage.kt`, `feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/pages/SettingsPage.kt`, and `feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountScreen.kt`.

Also applied minor UX refinements: ensured date picker text field and dropdowns follow the updated spacing, preserved accessibility content descriptions, and kept existing validation and view-model flows intact (`feature/recurringDeposit/src/commonMain/kotlin/com/mifos/feature/recurringDeposit/newRecurringDepositAccount/RecurringAccountViewModel.kt`). Verified visual alignment across steps, updated related localized strings where labels changed, and smoke-tested navigation and form submission paths to prevent regressions.

TheKalpeshPawar committed:
fix: correct variable naming and improve null safety in RecurringAccountViewModel and SettingsPage

TheKalpeshPawar committed:
fix: remove unused import and correct formatting in MifosOutlinedTextField

biplab1 commented on version.txt:
```diff
@@ -1 +1 @@
-2025.10.4-beta.0.6
\ No newline at end of file
+2025.11.2-beta.0.21
```
Revert this change as follows:
```suggestion
2025.10.4-beta.0.6
```

TheKalpeshPawar replied:
Done

TheKalpeshPawar committed:
Update version.txt

Co-authored-by: Biplab Dutta <biplabdutta27@gmail.com>

TheKalpeshPawar committed:
Merge branch 'development' into settings-recurring-deposit-account

biplab1 commented on core/data/src/commonMain/kotlin/com/mifos/core/data/repository/RecurringAccountRepository.kt:
```diff
@@ -17,9 +17,9 @@ import kotlinx.coroutines.flow.Flow
+    fun getRecurringAccountTemplate(): Flow<DataState<RecurringDepositAccountTemplate>>
 
-    fun getRecuttingAccountRepositoryBtProduct(
+    fun getRecurringAccountTemplateByProduct(
```
Both functions have the same API endpoints,  consider a single method with optional `productId` field.

I understand that this change is not a part of this PR, however, this should have been handled in your previous PR on API implementation itself. From your previous PR on API:

```kotlin
    @GET(APIEndPoint.CREATE_RECURRING_DEPOSIT_ACCOUNTS + "/template")
    fun getRecurringDepositAccountTemplate(): Flow<RecurringDepositAccountTemplate>

    @GET(APIEndPoint.CREATE_RECURRING_DEPOSIT_ACCOUNTS + "/template")
    fun getRecurringDepositAccountTemplateByProduct(
        @Query("clientId") clientId: Int,
        @Query("productId") productId: Int,
    ): Flow<RecurringDepositAccountTemplate>
```

biplab1 commented on core/data/src/commonMain/kotlin/com/mifos/core/data/repositoryImp/RecurringAccountRepositoryImp.kt:
```diff
@@ -21,12 +21,12 @@ import kotlinx.coroutines.flow.Flow
+    override fun getRecurringAccountTemplateByProduct(
         clientId: Int,
         productId: Int,
     ): Flow<DataState<RecurringDepositAccountTemplate>> {
```
Same here, please review.

TheKalpeshPawar replied:
yes sir. I have fixed the issue. I will push the commit once the project builds.

TheKalpeshPawar committed:
refactor: unify recurring account template retrieval methods and improve null safety

TheKalpeshPawar committed:
fix: enforce non-null clientId in recurring account template methods

biplab1 reviewed (approved):
This looks good to me. It can be merged

TheKalpeshPawar committed:
Merge branch 'development' into settings-recurring-deposit-account
